### DISCUSSION TITLE
Implementation for z/OS (IBM mainframe)

### DISCUSSION THREAD

**Discussion by andrew890** [PlummersSoftwareLLC/Primes#346] [General]
I was interested to try this on z/OS i.e. IBM mainframe operating system, using both C++ and Java.

C++ required a few changes due to newer compiler features not present.
Java ran without change, however a .gitattributes file with
```
*.java zos-working-tree-encoding=ibm-1047
```
was required to checkout and compile the Java code on z/OS.

Are these changes worth contributing back to the original repository? The C++ version will run under Linux, but there's nothing new and no performance improvement in the code itself. Just the ability to run on a different platform.

**Comment by rbergen** [comment]
I'll start with this question: 

> Are these changes worth contributing back to the original repository?

In my opinion. no. This repo targets the x86-64 and aarch64 architectures (the latter to the extent possible), and the existing C++ and Java solutions work on both. As I also mentioned in response to [this PR of yesterday](https://github.com/PlummersSoftwareLLC/Primes/pull/333), I would love to take a look at the details in a "show and tell" discussion post, though. 

With that said, I do want to express that I think it's wonderful that this project is reaching direct descendants of systems from the 1960s, as illustrated by the aforementioned PR and this discussion.

**Comment by davepl** [comment]
I, for one, would LOVE to see results on big iron!   If it can't run the entire docker suite, even a manual recompile of the C++ and Java versions would be cool just to get the numbers.  If you do the primes_par.cpp solution I can put it in context of comparing it to several other hardware platforms.

**Comment by rbergen** [comment]
If it is possible to run a few solutions on z/OS (again), then I think it would be great to also include the COBOL implementation included in #333. That language is of course period correct for the early ancestors of the mainframes that now run z/OS.

**Comment by andrew890** [comment]
The numbers on z/OS are difficult to compare to other systems (even other z/OS systems) due to variability. Due to the way software pricing has evolved, z/OS system capacity is VERY granular. The difference between the fastest CPU i.e. single thread speed and slowest is around 20x. Then you can have from 1-190 cores.

On my system (single core VM guest) I get around 4800 passes. The performance relative to other platforms seemed to improve for larger numbers e.g. 1 billion, which might be due to larger CPU cache. I can't do any meaningful parallel testing due to the single core.

The main interest for me is to compare different languages on z/OS. I don't know COBOL, but #333 looks easy enough to run so I can do that and compare to C++ and Java.

I am planning to write up my results and I would like to make the code available for other z/OS people to try. I can do that in my own repository with a link back to the original, or contribute it here.

Maybe a show and tell directory could be created in the original repository for things that don't fit the main project? Another candidate would be the C solution that does the calculation at compile time - it's very interesting, but not proper drag racing to start at the finish line!

**Reply by rbergen** [reply]
> I am planning to write up my results and I would like to make the code available for other z/OS people to try. I can do that in my own repository with a link back to the original, or contribute it here.

I think creating your own repository and sharing it in a "show & tell" post (that is an actual Discussions category) here would be awesome.

> Maybe a show and tell directory could be created in the original repository for things that don't fit the main project? Another candidate would be the C solution that does the calculation at compile time - it's very interesting, but not proper drag racing to start at the finish line!

This repository's contents must target the amd64 and/or arm64 architectures (preferably both), because that is a requirement for inclusion in the benchmark runs. Also, at this time we can only accept new submissions if they either add a new language, improve the performance of an existing solution, or demonstrate the performance impact of a specific implementation approach (usually leading to a change in one of the [key characteristics](https://github.com/PlummersSoftwareLLC/Primes/blob/drag-race/CONTRIBUTING.md#characteristics) compared to other entries).
The C++ entry you are referring to is one of the latter kind. Due the fact it does "start at the finish line" as you call it, it is _not_ an "official" entry for the actual drag race (it's marked as "unfaithful" for that reason).

As I said, I would love to have a Discussions post here that shares your work and results. I, for one, can't wait to see what they are.
